### PR TITLE
Player/02 implementing player

### PR DESCRIPTION
**Pull request by AleksKors**
### Player
  - Implemented Player
    - Added Player movement
    - Added Player aiming
    - Added Player health & collision

**Missing**
- Testing
- Player Inventory
- Player building
- Player customization

### CODE REVIEW

File: Player/pom.xml [hunk 1]
```diff
@@ -11,6 +11,12 @@
 	<artifactId>Player</artifactId>
 
 	<dependencies>
+		<dependency>
+			<groupId>${project.groupId}</groupId>
+			<artifactId>Common</artifactId>
+			<version>${project.version}</version>
+			<scope>compile</scope>
```
Comment (FearfulSnow) [inline Player/pom.xml:18]: remove this

File: Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+public class PlayerControl implements IEntityProcessingService {
+	private double walkSpeed = 10;
+	private double diagonalWalkSpeed = Math.sqrt(2 * (this.walkSpeed * this.walkSpeed));
+
+	@Override
+	public void process(IEntityManager entityManager, IGameSettings gameSettings) {
+		for (CommonEntity player : entityManager.getEntitiesByClass(Player.class)) {
+			//Set rotation to look a cursor
+			player.getCoordinates().getAngleBetween(gameSettings.getKeys().getMouseCoordinates());
```
Comment (FearfulSnow) [inline Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java:17]: does not set rotation

File: Player/src/main/java/com/github/sef24sp4/player/PlayerPlugin.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package com.github.sef24sp4.player;
+
+import com.github.sef24sp4.common.data.Coordinates;
+import com.github.sef24sp4.common.entities.CommonEntity;
+import com.github.sef24sp4.common.interfaces.IEntityManager;
+import com.github.sef24sp4.common.interfaces.IGameSettings;
+import com.github.sef24sp4.common.services.IGamePluginService;
+
+public class PlayerPlugin implements IGamePluginService {
+	private CommonEntity player;
```
Comment (FearfulSnow) [inline Player/src/main/java/com/github/sef24sp4/player/PlayerPlugin.java:10]: what not the player class

Comment (FearfulSnow) [reply Player/src/main/java/com/github/sef24sp4/player/PlayerPlugin.java:10]: why*

File: Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+			//TODO;
+			//Check if it should shoot
+			if (gameSettings.getKeys().isDown(InputAction.SHOOT)) {
+				System.out.println("Shooting not implemented yet");
+				//Shoot code here
+			}
+			//Player movement
+			if (gameSettings.getKeys().isDown(InputAction.UP) && gameSettings.getKeys().isDown(InputAction.LEFT)) {
+				player.setX(player.getX() - this.diagonalWalkSpeed);
```
Comment (MidnightRocket) [inline Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java:26]: You can extract `player.getX()` and `player.getY()` into variable.

File: Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+				player.setX(player.getX() + this.walkSpeed);
+			}
+			if (gameSettings.getKeys().isDown(InputAction.UP) && !gameSettings.getKeys().isDown(InputAction.LEFT)
+					&& !gameSettings.getKeys().isDown(InputAction.RIGHT)) {
+				player.setY(player.getY() + this.walkSpeed);
+			}
+			if (gameSettings.getKeys().isDown(InputAction.DOWN) && !gameSettings.getKeys().isDown(InputAction.LEFT)
+					&& !gameSettings.getKeys().isDown(InputAction.RIGHT)) {
+				player.setY(player.getY() - this.walkSpeed);
+			}
```
Comment (MidnightRocket) [inline Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java:56]: Should be an if-else chain instead.

File: Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+			if (player.getX() > gameSettings.getDisplayWidth()) {
+				player.setX(gameSettings.getDisplayWidth() - 1);
+			}
+			if (player.getY() < 0) {
+				player.setY(1);
+			}
+			if (player.getY() > gameSettings.getDisplayHeight()) {
+				player.setY(gameSettings.getDisplayHeight() - 1);
+			}
+		}
```
Comment (MidnightRocket) [inline Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java:70]: Can use the `gameSettings.isEntityWithinFrame(IEntity entity)` method.

File: Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import com.github.sef24sp4.common.interfaces.IGameSettings;
+import com.github.sef24sp4.common.services.IEntityProcessingService;
+
+public class PlayerControl implements IEntityProcessingService {
+	private double walkSpeed = 10;
+	private double diagonalWalkSpeed = Math.sqrt(2 * (this.walkSpeed * this.walkSpeed));
+
+	@Override
+	public void process(IEntityManager entityManager, IGameSettings gameSettings) {
+		for (CommonEntity player : entityManager.getEntitiesByClass(Player.class)) {
```
Comment (MidnightRocket) [inline Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java:15]: Make use of singleton player

File: Player/src/main/java/com/github/sef24sp4/player/PlayerPlugin.java
Comment (MidnightRocket) [inline Player/src/main/java/com/github/sef24sp4/player/PlayerPlugin.java:1]: Use some sort of singletone for Player

**Review by MidnightRocket — CHANGES_REQUESTED** [review]

**Comment by MidnightRocket** [comment]
Should wait until this https://github.com/SE-F24-SP4/semesterproject-4/issues/68 is merged into main and then integrated into Player.

File: Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+			}
+			if (playerX > gameSettings.getDisplayWidth()) {
+				player.setX(gameSettings.getDisplayWidth());
+			}
+			if (playerY < 0) {
+				player.setY(0);
+			}
+			if (playerY > gameSettings.getDisplayHeight()) {
+				player.setY(gameSettings.getDisplayHeight());
+			}
```
[the remark this answers is not in the document]
Comment (AleksKors) [reply Player/src/main/java/com/github/sef24sp4/player/PlayerControl.java:71]: Not sure if I understand the request. Can you elaborate?

File: Player/src/main/java/com/github/sef24sp4/player/Player.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	}
+	/**
+	 * The amount the entity should move
+	 * in the X and Y cords per game tick.
+	 * @return The walk speed
+	 */
+	public double getWalkSpeed() {
+		return this.walkSpeed;
+	}
+	public void setWalkSpeed(double speed) {
```
Comment (FearfulSnow) [inline Player/src/main/java/com/github/sef24sp4/player/Player.java:44]: should i put negative numbers here

File: Player/src/main/java/com/github/sef24sp4/player/Player.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+	@Override
+	public IGameMetadata getMetadata() {
+		return this.metadata;
+	}
+
+	@Override
+	public GameElementType getType() {
+		return this.metadata.getType();
+	}
```
[the remark this answers is not in the document]
Comment (FearfulSnow) [reply Player/src/main/java/com/github/sef24sp4/player/Player.java:72]: this is not resolved

File: Player/src/main/java/com/github/sef24sp4/player/Player.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		return this.maxHealth;
+	}
+	/**
+	 * Takes the entity's health and subtracts the damage.
+	 *
+	 * @param damage The damage of the attacking entity.
+	 */
+	public void takeDamage(double damage) {
+		this.health -= damage;
+	}
```
Comment (FearfulSnow) [inline Player/src/main/java/com/github/sef24sp4/player/Player.java:88]: can take in negative numbers

**Review by FearfulSnow — CHANGES_REQUESTED** [review]

File: Player/src/main/java/com/github/sef24sp4/player/Player.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+	/**
+	 * Sets the players movement speed to "speed".
+	 *
+	 * @param speed The amount the player should move per game tick. Needs to be positive.
+	 */
+	public void setWalkSpeed(double speed) {
+		try {
+			if (speed < 0) {
```
Comment (FearfulSnow) [inline Player/src/main/java/com/github/sef24sp4/player/Player.java:52]: you might need to make this: speed <= 0. if we want the player to stand still.

**Review by FearfulSnow — CHANGES_REQUESTED** [review]

**Review by FearfulSnow — APPROVED** [review]

**Review by MidnightRocket — APPROVED** [review]

**Pull request merged**